mutex.go

```diff
@@ -44,10 +44,12 @@ var (
 // Lock waits up to duration d for l.Lock() to succeed.
 func Lock(l sync.Locker, d time.Duration) error {
 	done := make(chan struct{}, 1)
-	go func() {
+	ctx, cancel := context.WithTimeout(context.Background(), d*time.Second)
+	defer cancel()
+	go func(context.Context) {
 		l.Lock()
 		done <- struct{}{}
-	}()
+	}(ctx)
```

Comment:
> simply passing a context that had been initialized with a timeout would cause the goroutine to timeout, after the requested interval.

That's magic! Go doesn't do magic! :)